README
GM2Calc
=======

GM2Calc calculates the SUSY contributions to the anomalous magnetic
moment of the muon a_mu = (g-2)/2 in the real MSSM at the 1- and
leading 2-loop level.

Homepage:                https://gm2calc.hepforge.org
Source code repository:  https://github.com/gm2calc
References:              arxiv:1510.08071


Requirements
============

 * C++ compiler (g++ >= 4.6.3 or clang++ >= 3.4 or icpc >= 14.0.3)
 * Boost (version 1.37.0 or higher)   http://www.boost.org

Optional:

 * LAPACK                             http://www.netlib.org/lapack/


Compilation
===========

To compile GM2Calc run GNU `make':

  $ make

The used compiler and include paths can be passed as an argument to
GNU `make'.  For example:

  $ make \
       CXX=icpc \
       EIGENFLAGS="-I/opt/eigen3/eigen3" \
       BOOSTFLAGS="-I/opt/boost/include"

The following default values are set in the makefile:

  CXX             := g++
  CPPFLAGS        := -Isrc
  CXXFLAGS        := -O2 -std=c++11 -fPIC
  CXX_DEP_GEN     := g++
  MAKELIB         := ar cru
  MAKESHAREDLIB   := $(CXX) -shared
  BOOSTFLAGS      := -I/usr/include
  EIGENFLAGS      := -I.

The command line arguments passed to `make' are stored in the file
`make.args'.  To (re-)run `make' with the command line arguments
stored in `make.args' run

  $ cat make.args | xargs make


Compilation of the example programs
-----------------------------------

GM2Calc ships two C++ example programs:

  examples/example-gm2calc.cpp
  examples/example-slha.cpp

and two C example programs:

  examples/example-gm2calc_c.c
  examples/example-slha_c.c

These example programs illustrate how to use routines of GM2Calc at
the C/C++ level.  The examples can be compiled by running

  $ make examples

For the compilation of the C programs a C compiler is needed.  The
used C compiler, the compiler flags and the linked libraries can be
customized as:

  $ make examples \
       CC=icc CFLAGS="-O2 -std=c99" CLIBS="-lstdc++ -lm" LDLIBS="-lirc"


Using LAPACK (optional)
-----------------------

The routines which diagonalize the mass matrices by default use the
algorithms provided by Eigen.  However, the routines provided by
LAPACK can be used alternatively.  In order to make use of the LAPACK
routines one has to pass `ENABLE_LAPACK=1' to the make command.  In
addition, the LAPACK and BLAS libraries to be linked have to be passed
in the `LDFLAGS' variable to the make command:

  $ make ENABLE_LAPACK=1 LDLIBS="-llapack -lblas"


Running GM2Calc
===============

GM2Calc can be run with (i) either an SLHA interface or with (ii) a
custom GM2Calc interface (similar to SLHA, but different definition of
input parameters).  See `bin/gm2calc.x --help' for all options.

Example using the SLHA interface:

  $ bin/gm2calc.x --slha-input-file=input/example.slha

or

  $ cat input/example.slha | bin/gm2calc.x --slha-input-file=-

Example using the GM2Calc interface:

  $ bin/gm2calc.x --gm2calc-input-file=input/example.gm2

or

  $ cat input/example.gm2 | bin/gm2calc.x --gm2calc-input-file=-


SLHA interface
==============

When the input parameters are provided in SLHA-compliant form, GM2Calc
reads the input parameters from the following blocks:

 * SMINPUTS: Standard Model fermion masses, W and Z pole masses,
   alpha_s(MZ)

 * MASS: pole masses of SUSY particles, W pole mass.  If the W pole
   mass is given in MASS[24], then this value will be used instead of
   the W pole mass given in SMINPUTS[9].

 * HMIX: tan(beta), mu parameter, renormalization scale

 * AU, AD, AE: soft-breaking trilinear couplings, renormalization
   scale

 * MSOFT: soft-breaking gaugino masses, soft-breaking sfermion masses,
   renormalization scale

 * GM2CalcInput: alpha_em(MZ), alpha_em in the Thomson limit


GM2Calc interface
=================

When the input parameters are provided in GM2Calc specific form,
GM2Calc reads the input parameters from the following blocks:

 * SMINPUTS: Standard Model fermion masses, W and Z pole masses,
   alpha_s(MZ)

 * GM2CalcInput: renormalization scale, alpha_em(MZ), alpha_em in the
   Thomson limit, tan(beta), mu parameter, soft-breaking gaugino
   masses, soft-breaking Bmu parameter, soft-breaking sfermion masses,
   soft-breaking trilinear couplings


C/C++ interface
===============

GM2Calc provides a C and a C++ interface.  To use the routines of
GM2Calc in a C++ program, the following C++ header files have to be
included:

   src/gm2_1loop.hpp
   src/gm2_2loop.hpp
   src/gm2_uncertainty.hpp
   src/MSSMNoFV_onshell.hpp

To use the routines of GM2Calc in a C program, the following C header
files have to be included:

   src/gm2_1loop.h
   src/gm2_2loop.h
   src/gm2_uncertainty.h
   src/MSSMNoFV_onshell.h

Please refer to the content of these header files for a precise
definition of all interface functions.  The C/C++ example programs
also serve as an illustration of the interface routines.


Mathematica interface
=====================

In order to use GM2Calc from within Mathematica, first create
GM2Calc's MathLink executable by running

  $ make mathlink

By default the MathLink compiler is the bin/mcc script, which finds
the location of the MathLink compiler for most UNIX-like environments
using the Mathematica kernel.  The Mathematica kernel can be given via
the MATH variable.  Alternatively, the MathLink compiler can be set
explicitly via the MCC variable.

Example:

  $ make \
       CXX=icpc CC=icc \
       LDLIBS="-lboost_math_c99" \
       MATH=math10.1

The MathLink executable can be found in

  bin/gm2calc.mx

This executable can be installed in Mathematica by calling

  Install["bin/gm2calc.mx"]

Afterwards, the GM2Calc Mathematica interface functions can be used,
see examples/example-slha.m and examples/example-gm2calc.m .


Configuration options
=====================

The calculation of a_mu can be controlled in the GM2CalcConfig block.

 * GM2CalcConfig[0]: output format (0 ... 4)

   0: minimal output (a single number)
   1: detailed (a detailed output of the various contributions)
   2: write the value of (g-2)/2 into LOWEN block, entry 6
   3: write the value of (g-2)/2 into SPhenoLowEnergy block, entry 21
   4: write the value of (g-2)/2 into GM2CalcOutput block, entry 0

 * GM2CalcConfig[1]: loop order of the calculation (0, 1 or 2)

 * GM2CalcConfig[2]: disable/enable tan(beta) resummation (0 or 1)

 * GM2CalcConfig[3]: force output even if physical problem has occured
   (0 or 1).  WARNING: The result might not be trusted if a problem
   has occured!

 * GM2CalcConfig[4]: disable/enable verbose output (0 or 1).

 * GM2CalcConfig[5]: disable/enable uncertainty estimation (0 or 1).

   Depending on the chosen output format, GM2CalcConfig[0], the
   uncertainty is written

   as a sigle number to stdout    in case of minimal output,
   to the first line              in case of detailed output,
   to GM2CalcOutput[1]            otherwise.


Source code documentation
=========================

GM2Calc uses Doxygen to document the source code.  The complete source
code documentation can be generated by running

  $ make doc

Afterwards, the generated HTML pages can be found in doc/html/ and can
be openend with your favourite web browser, e.g.

  $ firefox doc/html/index.html


ChangeLog
GM2Calc-1.3.3 [July, 19 2017]

 * Optimization (commit aa7afc0): Avoid redundant calls to complicated
   lambda_mu_cha() function.

 * Bugfix (commit cbb4df2): Fix compilation error on Cygwin where M_PI
   might not be defined in <math.h>.

 * Bugfix (commits 9d4f768, d820a53, b9a6320, 54970a6): Catch floating
   point overflow/underflow from diagonalization of mass matrices
   during DR-bar to on-shell conversion of right-handed smuon mass
   parameter.

 * Bugfix (commit 54b86e8): A small coefficient in the complex dilog
   has been corrected.  However, the complex dilog is not used so far.

GM2Calc-1.3.2 [February, 22 2017]

 * Bugfix (commit e8af0c9): Catch potential exceptions from C
   interface functions which calculate amu w/o tan(beta) resummation.

 * Bugfix (commit 654276e): Allow user to pass numeric values to
   GM2CalcAmuSLHAScheme[], GM2CalcAmuGM2CalcScheme[] and
   GM2CalcSetSMParameters[] which are not numbers, but which would
   evaluate to numbers.

GM2Calc-1.3.1 [January, 31 2017]

 * Bugfix (commit ab73c49): Workaround bug in mcc 11.0.0.

GM2Calc-1.3.0 [July, 21 2016]

 * Feature: A Mathematica interface for GM2Calc has been added.  To
   build it, run `make mathlink'.  The compiled MathLink executable
   can then be found in bin/gm2calc.mx .  To use it, the MathLink
   executable must be installed in Mathematica via

      Install["bin/gm2calc.mx"];

   Two examples using the Mathematica interface of GM2Calc can be found in

      examples/example-gm2calc.m
      examples/example-slha.m

   These two examples behave exactly like their C/C++ counterparts.

 * Change (commit b3c7357): Abort calculation if tan(beta) is
   undefined or zero.

 * Change (commit 910b1bd): Abort calculation if the mu parameter is
   zero.  For mu = 0 the approximate fermion/sfermion 2-loop
   corrections are ill-defined.

 * Change (commit 16b181b): Abort calculation if the lightest chargino
   mass is zero.  In that case the photonic 2-loop contribution
   originating from loop function F3C is ill-defined.

 * Change (commits 6d2a8b3, 2a8b2ed): Create shared library
   src/libgm2calc.so (in addition to the static library) for
   convenience.

 * Change (commit c0b5923): Compile all .cpp files in src/ and put all
   generated object files into the libraries, except for
   src/gm2calc.o.  In this way, there is no need for GAMBIT to modify
   src/module.mk to add further .cpp files for compilation.

 * Bugfix (commit b3d3c0f): Implementation of the x = 0 limits of the
   F1N[x], F2N[x], F3N[x], F4N[x] and F1C[x] functions.

 * Bugfix (commit 3ec5df4): Catch potential exception during iteration
   to determine the QCD scale, Lambda_QCD.  This exception is thrown,
   for example, if the Z pole mass is set to be larger than 5 TeV.

GM2Calc-1.2.0 [June, 21 2016]

 * Feature (commit b65d75d): Adding C interface functions to
   retrieving warnings and problems in form of C strings.

 * Feature (commit ffe9b50): Perform the calculation of the
   uncertainty of a_mu(0-loop) and a_mu(1-loop).  The uncertainty of
   a_mu(0-loop) is the magnitude of a_mu(1-loop).  The uncertainty of
   a_mu(1-loop) is the sum of the magnitudes of a_mu(2-loop,best) and
   the uncertainty of a_mu(2-loop,best).

 * Change (commit fbee2e3): Adding calculation of amu uncertainty to
   the C/C++ example programs.

 * Bugfix (commit 0665a09): Correcting the used sbottom masses in
   Delta_b corrections.

 * Bugfix (commit 71b0cf3): Do not allow the calculation of amu for
   negative soft-breaking squared sfermion mass parameters.

 * Bugfix (commit 7d4d4e7): Use better initial guess for the
   root-finding algorithm, which determines the soft-breaking squared
   mass parameter of the right-handed smuon from the mostly
   right-handed smuon pole mass.

 * Bugfix (commit 9eecf61): Catching potential exception during the
   DR-bar to on-shell conversion of the soft-breaking squared mass
   parameter of the right-handed smuon.

GM2Calc-1.1.2 [April, 08 2016]

 * Bugfix (commit a374bf7): Reformulate Delta_b corrections to avoid
   numerical problems when Mu, M3 or M1 are zero.

 * Bugfix (commit 4177460): Implement limits of Fa(), Fb() and
   amuBmuLmuR() functions when one or more masses go to zero.

GM2Calc-1.1.1 [March, 29 2016]

 * Bugfix (commit 9ca7890, 33e68a4, c047d46): Implement limit of
   Iabc() function when one or more masses go to zero.

GM2Calc-1.1.0 [December, 14 2015]

 * Change (commit 5cbbc50): The example programs have been moved to a
   separate examples/ directory.

 * Change (commit 873133f): Adding a C interface.  The new headers

      src/gm2_1loop.h
      src/gm2_2loop.h
      src/gm2_uncertainty.h
      src/MSSMNoFV_onshell.h

   declare C interface functions for GM2Calc routines.  In addition,
   two example C programs

      examples/example-gm2calc_c.c
      examples/example-slha_c.c

   have been added to illustrate the C interface.  These two C example
   programs behave exactly like their C++ counterparts.  They can be
   compiled by running `make examples'.

GM2Calc-1.0.0 [October, 29 2015]

 * Official release 1.0.0.

GM2Calc-0.2.17 [October, 26 2015]

 * Feature (commit 132851): The uncertainty of amu(2L,TB resummed) can
   be calculated by setting GM2CalcConfig[5] to 1.  Depending on the
   chosen output format (GM2CalcConfig[0]) the uncertainty is written

   as a single number to stdout   in case of minimal output,
   to the first line              in case of detailed output,
   to GM2CalcOutput[1]            in case of NMSSMTools output,
   to GM2CalcOutput[1]            in case of SPheno output,
   to GM2CalcOutput[1]            in case of GM2Calc output.

 * Feature (commit 02841d): A new SLHA-compliant output format
   "GM2Calc" has been added to avoid interference with SPheno or
   NMSSMTools: If GM2CalcConfig[0] is set to 4, the value of amu is
   written to GM2CalcOutput[0].  If uncertainty estimation has been
   enabled in addition, the uncertainty is written to
   GM2CalcOutput[1].

src/dilog.cpp
// ====================================================================
// This file is part of GM2Calc.
//
// GM2Calc is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published
// by the Free Software Foundation, either version 3 of the License,
// or (at your option) any later version.
//
// GM2Calc is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with GM2Calc.  If not, see
// <http://www.gnu.org/licenses/>.
// ====================================================================

#include "dilog.hpp"
#include <cmath>
#include <limits>

namespace gm2calc {

namespace {
   template <typename T>
   T sqr(T x) { return x*x; }
}

/**
 * @brief Real dilogarithm \f$\mathrm{Li}_2(z)\f$
 * @param x real argument
 * @note Implementation translated by R.Brun from CERNLIB DILOG function C332
 * @return \f$\mathrm{Li}_2(z)\f$
 */
double dilog(double x) {
   const double PI = 3.14159265358979323846;
   const double HF  = 0.5;
   const double PI2 = PI*PI;
   const double PI3 = PI2/3;
   const double PI6 = PI2/6;
   const double PI12 = PI2/12;
   const double C[20] = {0.42996693560813697, 0.40975987533077105,
     -0.01858843665014592, 0.00145751084062268,-0.00014304184442340,
      0.00001588415541880,-0.00000190784959387, 0.00000024195180854,
     -0.00000003193341274, 0.00000000434545063,-0.00000000060578480,
      0.00000000008612098,-0.00000000001244332, 0.00000000000182256,
     -0.00000000000027007, 0.00000000000004042,-0.00000000000000610,
      0.00000000000000093,-0.00000000000000014, 0.00000000000000002};

   double T,H,Y,S,A,ALFA,B1,B2,B0;

   if (x == 1) {
       H = PI6;
   } else if (x == -1) {
       H = -PI12;
   } else {
       T = -x;
       if (T <= -2) {
           Y = -1/(1+T);
           S = 1;
           B1= log(-T);
           B2= log(1+1/T);
           A = -PI3+HF*(B1*B1-B2*B2);
       } else if (T < -1) {
           Y = -1-T;
           S = -1;
           A = log(-T);
           A = -PI6+A*(A+log(1+1/T));
       } else if (T <= -0.5) {
           Y = -(1+T)/T;
           S = 1;
           A = log(-T);
           A = -PI6+A*(-HF*A+log(1+T));
       } else if (T < 0) {
           Y = -T/(1+T);
           S = -1;
           B1= log(1+T);
           A = HF*B1*B1;
       } else if (T <= 1) {
           Y = T;
           S = 1;
           A = 0;
       } else {
           Y = 1/T;
           S = -1;
           B1= log(T);
           A = PI6+HF*B1*B1;
       }
       H    = Y+Y-1;
       ALFA = H+H;
       B1   = 0;
       B2   = 0;
       for (int i=19;i>=0;i--){
          B0 = C[i] + ALFA*B1-B2;
          B2 = B1;
          B1 = B0;
       }
       H = -(S*(B0-H*B2)+A);
    }
    return H;
}

/**
 * @brief Complex dilogarithm \f$\mathrm{Li}_2(z)\f$
 * @param z complex argument
 * @note Implementation translated from SPheno to C++
 * @return \f$\mathrm{Li}_2(z)\f$
 */
std::complex<double> dilog(const std::complex<double>& z) {
   const double PI = 3.14159265358979323846;
   std::complex<double> cy, cz;
   int jsgn, ipi12;
   static const unsigned N = 20;

   // bf[1..N-1] are the even Bernoulli numbers / (2 n + 1)!
   // generated by: Table[BernoulliB[2 n]/(2 n + 1)!, {n, 1, 19}]
   const double bf[N] = {
      - 1./4.,
      + 1./36.,
      - 1./36.e2,
      + 1./21168.e1,
      - 1./108864.e2,
      + 1./52690176.e1,
      - 4.0647616451442255268059093862919666745470571274397078e-11,
      + 8.9216910204564525552179873167527488515142836130490451e-13,
      - 1.9939295860721075687236443477937897056306947496538801e-14,
      + 4.5189800296199181916504765528555932283968190144666184e-16,
      - 1.0356517612181247014483411542218656665960912381686505e-17,
      + 2.3952186210261867457402837430009803816789490019429743e-19,
      - 5.5817858743250093362830745056254199055670546676443981e-21,
      + 1.3091507554183212858123073991865923017498498387833038e-22,
      - 3.0874198024267402932422797648664624315955652561327457e-24,
      + 7.315975652702203420357905609252148591033401063690875e-26,
      - 1.7408456572340007409890551477597025453408414217542713e-27,
      + 4.1576356446138997196178996207752266734882541595115639e-29,
      - 9.9621484882846221031940067024558388498548600173944888e-31,
      + 2.3940344248961653005211679878937495629342791569329158e-32,
   };

   const double rz = std::real(z);
   const double iz = std::imag(z);
   const double az = std::sqrt(sqr(rz) + sqr(iz));

   // special cases
   if (iz == 0.) {
      if (rz <= 1.)
         return std::complex<double>(dilog(rz), 0.);
      if (rz > 1.)
         return std::complex<double>(dilog(rz), -PI*std::log(rz));
   } else if (az < std::numeric_limits<double>::epsilon()) {
      return z;
   }

   // transformation to |z|<1, Re(z)<=0.5
   if (rz <= 0.5) {
      if (az > 1.) {
         cy = -0.5 * sqr(std::log(-z));
         cz = -std::log(1. - 1. / z);
         jsgn = -1;
         ipi12 = -2;
      } else { // (az <= 1.)
         cy = 0;
         cz = -std::log(1. - z);
         jsgn = 1;
         ipi12 = 0;
      }
   } else { // rz > 0.5
      if (az <= std::sqrt(2*rz)) {
         cz = -std::log(z);
         cy = cz * std::log(1. - z);
         jsgn = -1;
         ipi12 = 2;
      } else { // (az > sqrt(2*rz))
         cy = -0.5 * sqr(std::log(-z));
         cz = -std::log(1. - 1. / z);
         jsgn = -1;
         ipi12 = -2;
      }
   }

   // the dilogarithm
   const std::complex<double> cz2(sqr(cz));
   std::complex<double> sumC;

   for (unsigned i1 = 2; i1 < N; i1++)
      sumC = cz2 * (sumC + bf[N + 1 - i1]);

   // lowest order terms w/ different powers
   sumC = cz + cz2 * (bf[0] + cz * (bf[1] + sumC));

   const std::complex<double> result
      = double(jsgn) * sumC + cy + ipi12 * PI * PI / 12.;

   return result;
}

} // namespace gm2calc
